src/psyclone/domain/common/transformations/kernel_module_inline_trans.py

```diff
@@ -0,0 +1,268 @@
+# Authors R. W. Ford, A. R. Porter, S. Siso and N. Nobre, STFC Daresbury Lab
+#         A. B. G. Chalk STFC Daresbury Lab
+#         J. Henrichs, Bureau of Meteorology
+# Modified I. Kavcic, Met Office
+
+''' This module provides the KernelModuleInlineTrans transformation. '''
+
+
+from psyclone.psyGen import Transformation, CodedKern
+from psyclone.psyir.transformations import TransformationError
+from psyclone.psyir.symbols import RoutineSymbol, DataSymbol, \
+    DataTypeSymbol, Symbol
+from psyclone.psyir.nodes import Container, ScopingNode, Reference, Routine, \
+    Literal, CodeBlock, Call
+
+
+class KernelModuleInlineTrans(Transformation):
+    ''' Module-inlines (bring the subroutine to the same compiler-unit) the
+    subroutine pointed by this Kernel. For example:
+
+    >>> invoke = ...
+    >>> schedule = invoke.schedule
+    >>>
+    >>> inline_trans = KernelModuleInlineTrans()
+    >>>
+    >>> inline_trans.apply(schedule.children[0].loop_body[0])
+    >>> # Uncomment the following line to see a text view of the schedule
+    >>> # print(schedule.view())
+
+    .. warning ::
+        Not all kernels subroutines can be module-inlined. This transformation
+        will reject attempts to in-line kernels that access global data in the
+        original module.
+
+    '''
+
+    def __str__(self):
+        return "Inline a kernel subroutine into the PSy module"
+
+    @property
+    def name(self):
+        ''' Returns the name of this transformation as a string.'''
+        return "KernelModuleInline"
+
+    def validate(self, node, options=None):
+        '''
+        Checks that the supplied node is a Kernel and that it is possible to
+        inline the PSyIR of its contents.
+
+        :param kern: the kernel which is the target of the transformation.
+        :type kern: :py:class:`psyclone.psyGen.CodedKern`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the target node is not a sub-class of \
+                                     psyGen.Kern.
+        :raises TransformationError: if the subroutine containing the \
+                                     implementation of the kernel cannot be \
+                                     retrieved wiht 'get_kernel_schedule'.
+        :raises TransformationError: if the kernel cannot be safely inlined.
+
+        '''
+
+        if not isinstance(node, CodedKern):
+            raise TransformationError(
+                f"Target of a {self.name} must be a sub-class of "
+                f"psyGen.CodedKern but got '{type(node).__name__}'")
+
+        # Check that the PSyIR and associated Symbol table of the Kernel is OK.
+        # If this kernel contains symbols that are not captured in the PSyIR
+        # SymbolTable then this raises an exception.
+        try:
+            kernel_schedule = node.get_kernel_schedule()
+        except Exception as error:
```

Comment:
Unless you disagree, I think catching Exception is the right thing to do here, since we say that we can recover from any problem in get_kernel_schedule  (by not inlining the node) and the reason will be displayed thanks to the raise-from syntax.

Also, if we are precise about the Exceptions that we know happen right now, we would need to remember every time that we add a new exception in the `get_kernel_schedule` codepath to update this. Add this would be quite hard to remember.

Reply:
OK, that sounds alright.